### PR TITLE
Autoconf update

### PR DESCRIPTION
**Pull request by rfalgout**
Update autoconf version to 2.71.  Since this changes `configure` considerably, test extensively before merging.  See issue #755 for more info on this update.

- [x] Check that the regression tests all run cleanly
  - [x] machine-tux
  - [x] machine-tux-spack
  - [x] machine-rztopaz
  - [x] machine-syrah
  - [x] machine-lassen
  - [x] machine-mac
- [ ] PETSc
- [x] WSC
- [x] MFEM

### CODE REVIEW

**Comment by rfalgout** [comment]
When I run `configure` on our Tux boxes, the option `-std=gnu11` is added to the compile line.  This does not happen on my macbook.  I assume it has something to do with the version of gcc I am using on the two systems?  Anybody know why this happens?  I just want to understand it.  The code seems to compile just fine with or without the option.

**Comment by rfalgout** [comment]
FYI, the openmp compile is failing and will require some investigation.

**Comment by rfalgout** [comment]
Another FYI, the following warning is generated by the `basic-debug-cpp` test in `machine-tux.sh`

    configure: WARNING: unknown Fortran name-mangling scheme

More investigation to do. :(

**Comment by rfalgout** [comment]
The tux regression tests are clean now except for the Fortran name-mangling warning message mentioned above.  I am trying to decide if there is something to actually fix there or if the warning should just be filtered out.  Let me know if you have any ideas/opinions here.

The issue with openmp mentioned before was a result of the fact that `configure` defines the C compiler, i.e. $CC, to be "mpicc -std=gnu11".  This broke the following line in the previous code because of the quotes that were around `${CC}`.

https://github.com/hypre-space/hypre/blob/2b69e00a035e1ef8614f9db8a4ba662ff7095b4f/src/config/hypre_macros_misc.m4#L101

Without the quotes, `basename` produces "mpicc" and switches correctly to set compiler options.  I'm not sure why the "-std=gnu11" is being added to begin with, but it appears to be related to the fact that the `gcc` compiler on my tux box is old, version 4.8.5.  I could not find a way around this.  If you have any ideas, let me know.

**Review by liruipeng — APPROVED** [review]
Looks good to me.

File: src/config/configure.in
Comment (liruipeng) [inline src/config/configure.in:2160]: Do we make the change because we only check one header-file (as oppose to `AC_CHECK_HEADERS` checking multiple)? Thanks

Comment (rfalgout) [reply src/config/configure.in:2160]: There was an error with 2.71 autoconf with the previous macro, but I don't remember exactly what the issue was.  I'm pretty sure the macro used in the fix is okay though.  I'll take another look at AC_CHECK_HEADERS.

Comment (rfalgout) [reply src/config/configure.in:2160]: Here is the error message:
````
tux406 > config/update.sh 
HYPRE version number is the same as the current version
configure.in:2143: warning: AC_CHECK_HEADERS("${CUDA_HOME}/include/cuda.h"): you should use literals
./lib/autoconf/headers.m4:217: AC_CHECK_HEADERS is expanded from...
configure.in:2143: the top level
configure.in:2147: warning: AC_CHECK_HEADERS("${CUDA_PATH}/include/cuda.h"): you should use literals
./lib/autoconf/headers.m4:217: AC_CHECK_HEADERS is expanded from...
````

File: src/config/HYPRE_config.h.in
Comment (waynemitchell) [inline src/config/HYPRE_config.h.in:30]: What's the reason for adding HAVE_STDIO_H and removing HAVE_MEMORY_H above? I'm not familiar with what these are actually doing anyway...

Comment (rfalgout) [reply src/config/HYPRE_config.h.in:30]: This happened automatically somehow as a result of running the new 2.71 version of autoconf.  Not sure why.

Comment (waynemitchell) [reply src/config/HYPRE_config.h.in:30]: Oh, I thought all the files with .in were inputs that we set by hand as opposed to auto-generated files. Good to know.

Comment (rfalgout) [reply src/config/HYPRE_config.h.in:30]: Generally they are, but sometimes autoconf will want to update these `.in` files as well.  It's a really confusing system.

Comment (rfalgout) [reply src/config/HYPRE_config.h.in:30]: Note that I also ran an "update" script (forgot the name, but it was recommended by autoconf when I was doing all this).  Some things were changed as a result of that script.  All I know is I did NOT manually add HAVE_STDIO_H myself. :)

Comment (liruipeng) [reply src/config/HYPRE_config.h.in:30]: Yep, it also got me. At the top it says `/* config/HYPRE_config.h.in. Generated from configure.in by autoheader. */`

Comment (prj-) [reply src/config/HYPRE_config.h.in:30]: > forgot the name, but it was recommended by autoconf when I was doing all this

Probably `autoupdate`?

**Comment by prj-** [comment]
PETSc pipeline is clean (https://gitlab.com/petsc/petsc/-/pipelines/685842688) with the hash 36f3242ba0adff4d704dd43530af335195901770.

**Comment by rfalgout** [comment]
> PETSc pipeline is clean (https://gitlab.com/petsc/petsc/-/pipelines/685842688) with the hash [36f3242](https://github.com/hypre-space/hypre/commit/36f3242ba0adff4d704dd43530af335195901770).

Thanks!  I may ask you to test this again, depending on how things go.

**Comment by rfalgout** [comment]
@liruipeng or @waynemitchell .  Would one of you try running the Lassen tests on this branch while I am gone?  I did a quick configure/compile on rztopaz, syrah, and lassen.  They all add '-std=gnu11' to the C compiler line.  I didn't realize I was using GNU compilers so extensively on these machines, and they all have a consistently old version number (4.9.3).

**Comment by liruipeng** [comment]
> @liruipeng or @waynemitchell . Would one of you try running the Lassen tests on this branch while I am gone? I did a quick configure/compile on rztopaz, syrah, and lassen. They all add '-std=gnu11' to the C compiler line. I didn't realize I was using GNU compilers so extensively on these machines, and they all have a consistently old version number (4.9.3).

Will do. Thanks!

**Comment by liruipeng** [comment]
> * machine-lassen

machine-lassen tests are all clean.

**Comment by prj-** [comment]
Could I help in any way get things moving?

**Comment by prj-** [comment]
Any update @rfalgout?

**Comment by rfalgout** [comment]
Hi @osborn9 .  I left you off of this originally because it was about our autoconf build system, but clearly you should be in the loop given that spack currently uses the autoconf version.  Thanks!

**Comment by rfalgout** [comment]
Hi @prj- .  Would you mind testing this branch again since there have been some changes since you last tested it?  We should be able to get this into `master` this week and do a release shortly after that.  Thanks!

**Comment by prj-** [comment]
That's great news, if all goes according to your schedule, that would be in time for the 3.19.0 release of PETSc. Here is the (partial) pipeline: https://gitlab.com/petsc/petsc/-/pipelines/777263716.

**Comment by rfalgout** [comment]
> That's great news, if all goes according to your schedule, that would be in time for the 3.19.0 release of PETSc. Here is the (partial) pipeline: https://gitlab.com/petsc/petsc/-/pipelines/777263716.

Yes, we are trying to do this in time for the PETSc release.  BTW, what would be a good target date for us to get a hypre release done?

**Comment by prj-** [comment]
Satish usually creates the tarball at the end of the month (I don't know whether this year will be any different, @balay will be able to comment on this). So if 2.27.1 or 2.28.0 gets released before the last week of March, that should be alright to get integrated.

**Comment by balay** [comment]
Yeah petsc-3.19 is targeted for end of march.

**Comment by rfalgout** [comment]
Hi All.  This PR is ready to merge from my perspective.  Just need to get a review from @prj- @waynemitchell @osborn9.  Also, let me know if we should test any other platforms.  Thanks!

**Review by osborn9 — APPROVED** [review]
Looks good to me! I've tested building via spack.

**Comment by prj-** [comment]
Cannot approve as a non-maintainer, but here is a thumb up 😅. I did not relaunch a partial PETSc pipeline but tested your latest commit on a local machine (which cannot use `master`) and it's fixed properly now with this branch. Also tried on Windows with MinGW (which is not covered by PETSc pipeline) and there does not seem to be any regression. Thank you

File: src/config/configure.in
Comment (waynemitchell) [inline src/config/configure.in:62]: Why 2.69 here? Don't we now want to require 2.71?

Comment (rfalgout) [reply src/config/configure.in:62]: I want to allow users to build the `configure` script with `autoconf` 2.69 in the event they have issues with this update.  I don't think there will be issues, but it's a fall back that helped to ease concerns some of our users had about this update.  I've tested everything to make sure 2.69 still produces a working `configure` script.

Comment (liruipeng) [reply src/config/configure.in:62]: So, 2.69 is the minimum version we support from now on?

Comment (prj-) [reply src/config/configure.in:62]: (which is 11-year old, BTW)

Comment (rfalgout) [reply src/config/configure.in:62]: > So, 2.69 is the minimum version we support from now on?

Yes, sort of.  I don't think we'll have to worry about this.  I am not aware of any users who have ever built their own `configure` script from our `configure.in`.  We don't want them to do that.  I just want to have a fall back for a short time, just in case somebody has an issue.  Developers will still be required to use 2.71.  Note that the numbers we had in the `AC_PREREQ` statements before were all over the map, but we always used 2.69 to generate the script.

Comment (rfalgout) [reply src/config/configure.in:62]: > (which is 11-year old, BTW)

Yes, it's old, but it's the default version of `autoconf` installed on all of the systems here at LLNL, so changing to 2.71 makes our users here nervous (and rightly so).

Comment (prj-) [reply src/config/configure.in:62]: Yes, it's the default on most of my systems as well (except my laptop), but requiring this as a minium, instead of 2.59 which is 20-year old should not impact that many people (if any).

Comment (waynemitchell) [reply src/config/configure.in:62]: Thanks for the explanation. All this looks good to me. Thanks!

Comment (balay) [reply src/config/configure.in:62]: CentOS9 [aka RHEL9] has 2.69 - so its likely to live for another 4-9 years?..

ref: https://access.redhat.com/product-life-cycles/?product=Red%20Hat%20Enterprise%20Linux,OpenShift%20Container%20Platform%204

**Review by waynemitchell — APPROVED** [review]

File: src/config/configure.in
Comment (prj-) [reply src/config/configure.in:62]: As long as Intel (or any other vendor) does not screw things up again, there should be no need for another update.

**Pull request merged**
